src/test/ui/did_you_mean/issue-31424.stderr

```diff
@@ -1,6 +1,8 @@
 error[E0596]: cannot borrow immutable argument `self` as mutable
   --> $DIR/issue-31424.rs:17:15
    |
+16 |     fn foo(&mut self) {
+   |            --------- given this existing mutable borrow...
```

Comment:
I agree the existing message is confusing, but this doesn't seem *quite* right. The problem is that `&mut self` desugars to `self: &mut Self`, and hence the variable `self` itself is immutable and can't be borrowed mutably. I find pointing out the "existing mutable borrow" somewhat misleading, since it suggests that this is a problem of a borrow conflicting with another action, but that's not right right.

For example, [this code compiles](https://play.rust-lang.org/?gist=74149b50b7821141edd675a786a82a69&version=stable):

```rust
// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
// file at the top-level directory of this distribution and at
// http://rust-lang.org/COPYRIGHT.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

struct Foo {
    u :u32,
}
impl Foo {
    fn foo(mut self: &mut Self) {
        || {
            let r = &mut self;
            r.u += 1;
        };
    }
    //~^^^^^ ERROR closure cannot assign to immutable argument `self` [E0595]
}

fn main() {}
```

I have to wonder  -- rather than improve the diagnostics here, maybe we should make `&mut self` desugar to `mut self: &mut Self` so that these code snippets work?

Reply:
I agree that `&mut self` should desugar to `mut self` and should probably be accepted. That being said, I would like to keep the label to other cases like `TestEnum::Item(ref mut x)`.

Reply:
I wonder if they should all desugar the same way? It seems like the error message is equally misleading here (or at least I find it misleading). The problem is that the binding is not mutable -- but *why* isn't it, anyway? I feel like we've been moving towards a place where it's ok to `&mut` borrow an `&mut` -- thanks to deref coercions etc -- and I wonder if we should just embrace it. 

Not sure how to make this decision. Feels like a thin *RFC*, but still a highly visible change in some sense, certainly would want **at least** FCP!

cc @rust-lang/lang -- thoughts?  The context is that `&mut self` and `ref mut foo` both desugar to an immutable binding, but people sometimes stub their toe writing `&mut self` or `&mut foo`, which results in an error (where `&mut *self` and `&mut *foo` would work).

@estebank -- that reminds me, the suggestion to remove `&mut` (instead of adding `*`) is not necessarily a good one. Removing `&mut` can result in moves, where adding `*` would work more often and be closer to the existing "intent" of the code, I think.

Reply:
Changing the semantics to accept this in all cases sounds eminently reasonable to me. I know I found it confusing at first, even though I'm now used to it.

Reply:
Interestingly, this also resolved something with match desugarings. I kind of forget the details now, but there was a minor inconsistency that centered about `ref mut` not being `mut`... maybe @cramertj remembers?